internal/biome/replay/record.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2020 YourBase Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package replay
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/yourbase/yb/internal/biome"
+)
+
+// A Recorder wraps another biome and records its commands in memory.
+// The interactions are saved to the filesystem when it is finished.
+type Recorder struct {
+	biome biome.Biome
+	dir   string
+
+	mu   sync.Mutex
+	data *replayData
+}
+
+// NewRecorder returns a new Recorder biome that wraps the given biome and
+// saves to dir on Close.
+func NewRecorder(dir string, bio biome.Biome) *Recorder {
+	return &Recorder{
+		biome: bio,
+		dir:   dir,
+		data: &replayData{
+			Descriptor:   bio.Describe(),
+			Dirs:         bio.Dirs(),
+			CleanedPaths: make(map[string]string),
+			AbsPaths:     make(map[string]bool),
+		},
+	}
+}
+
+// Close writes the recorded invocations to the directory passed into NewRecorder.
+func (rec *Recorder) Close() error {
+	rec.mu.Lock()
+	fname := recordFilename(rec.data.Descriptor)
+	jsonData, err := json.MarshalIndent(rec.data, "", "\t")
+	rec.mu.Unlock()
+	if err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	if err := os.MkdirAll(rec.dir, 0777); err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(rec.dir, fname), jsonData, 0666); err != nil {
+		return fmt.Errorf("close recorder: %w", err)
+	}
+	return nil
+}
+
+func recordFilename(desc *biome.Descriptor) string {
+	return desc.OS + "_" + desc.Arch + ".json"
+}
+
+func (rec *Recorder) Describe() *biome.Descriptor {
+	rec.mu.Lock()
+	defer rec.mu.Unlock()
+	return rec.data.Descriptor
+}
```

Comment:
You're right: they're not actually providing necessary synchronization.

I wasn't thinking about it deeply because the rest of the `replayData` struct is mutable, but those two fields stay constant. Since this isn't an obvious property, I added a comment to the mutex to indicate which fields of `replayData` it is protecting.